### DISCUSSION TITLE
Welcome to RevitTest Discussions!

### DISCUSSION THREAD

**Discussion by ricaun** [ricaun-io/RevitTest#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by Islam-Alqalyubi** [comment]
Thanks @ricaun for sharing this amazing work, I am really surprised by this work and I believe it is the first framework published for free to execute tests inside Revit. I am working to develop Revit plugins. I want to ask you about the skills, and courses I should learn to be able to develop testing framework like this amazing one.

**Reply by ricaun** [reply]
> Thanks @ricaun for sharing this amazing work, I am really surprised by this work and I believe it is the first framework published for free to execute tests inside Revit.

Thank you I appreciate it! I am not aware of a paid one, but have some similar projects about running tests inside Revit.

* Dynamo's Revit Tester Framework - https://github.com/DynamoDS/RevitTestFramework
* Geberit's Revit Test Runner - https://github.com/geberit/Revit.TestRunner
* Speckle's Revit Test Runner - https://github.com/specklesystems/xUnitRevit
* NeVeSpl's RevitTestLibrary - https://github.com/NeVeSpl/RevitTestLibrary

I used the same Revit objects injection in the tests as in Geberit`s project, at least was the test project at the time that was easy to make work. But, I have some specific requirements like running tests in Revit for desktop and/or in the Design Automation for Revit, and I tried to make work with Visual Studio, and worked, so that became a requirement. Making it convenient to run tests 😀

**Reply by ricaun** [reply]
> I am working to develop Revit plugins. I want to ask you about the skills, and courses I should learn to be able to develop testing framework like this amazing one.

Neat! Developing Revit plugins requires some skill and Revit has some specificity that could be a challenge. I'm not sure if there is a specific course, but always start with a question. How can I create an installation for a Revit plugin? How to sign a DLL to remove the popup when a plugin is loaded? How to create a library/dll to use in a plugin? How to make that library a package and post it on Nuget to reuse in another plugin? How to make two processes communicate with each other (Inter-Process Communication (IPC))? How to create the repository on GitHub? 

Anyway, I have a lot of private projects on my GitHub with the goal of testing and learning something new. This project has a lot of pieces of other projects, and I am pretty sure I over-engineer some stuff.

**Reply by ricaun** [reply]
I gonna release a new version tomorrow, I found some random issues that I didn't catch before the Revit 2025 release.

**Comment by pausie** [comment]
Hello @ricaun,
I'm encountering a MethodAccessException when trying to access certain methods from my test project. Despite making the methods public and ensuring that my class visibility settings are correctly configured, I continue to face this issue.

Environment:
NUnit Version: 3.13.3
ricaun.RevitTest.TestAdapter Version: 1.3.1
Revit Version: 2024
Development Environment: Visual Studio 2022
.NET Framework Targeted: 4.8

Error Message:
`System.MethodAccessException: Attempt by method 'Param.Tests.Tests.Test_CollectAllSpaces()' to access method 'Param.SmokeHandler.CollectAllSpaces(Autodesk.Revit.DB.Document)' failed.
`

Here's the test method where the exception occurs:
```
public void Test_CollectAllSpaces()
{
    var handler = new Param.SmokeHandler();

    // Arrange
    Document doc = uiapp.ActiveUIDocument.Document;

    // Act
    var spaces = handler.CollectAllSpaces(doc);

    // Assert
    Assert.IsNotNull(spaces, "Spaces list should not be null.");
    Assert.IsTrue(spaces.Count > 0, "There should be one or more spaces collected.");
}
```

And the method being tested:
```
public List<Space> CollectAllSpaces(Document doc)
 {
     List<Space> allSpaces = new List<Space>();

     // Samle lokale spaces
     var spaceCollector = new FilteredElementCollector(doc)
         .OfCategory(BuiltInCategory.OST_MEPSpaces)
         .WhereElementIsNotElementType()
         .OfType<Space>()
         .ToList();

     allSpaces.AddRange(spaceCollector);

     return allSpaces;
}
```

Could there be specific settings or configurations within the ricaun.RevitTest.TestAdapter that I might need to adjust to resolve this issue?Any help or insights you could provide would be greatly appreciated.

Thank you! :)

**Reply by ricaun** [reply]
That's strange... This class Param.SmokeHandler is a different dll or is the same as the test?

If is a different dll and you already have that DLL loaded inside Revit and in that DLL is private this message would happen I guess.

I don't know if is the case. But that should explain the issue.

**Reply by bimwood** [reply]
Yes, at first all my tests failed with 'System.MissingMethodException: Method not found' when accessing methods on classes that are in other DLLs in my solution. As you suggest, the tests work fine if I temporarily prevent Revit from loading my add-in.

**Reply by ricaun** [reply]
Usually is better to set to `internal` instead of `public` in the method you want to test. 
And in the csproj you can enable some assembly to find and visualize the internal methods/properties using the `InternalsVisibleTo`.

```xml
<ItemGroup>
  <InternalsVisibleTo Include="$(AssemblyName).Tests" />
</ItemGroup>
```
This is a really genetic code but if I add that in my csproj project `Sample` with the assembly name as `Sample`. And I have a test project with the name `Sample.Tests` this test project can visualize all the internal methods/properties in the `Sample` reference.

**Comment by MadhuByagari39** [comment]
Hi @ricaun , I'm adding RevitTest.testadapter to my VS 2022 Nunit project along with Revit All versions API x64 nuget to launch Revit 2025 when running test case from Test explorer. After modifying the Revit version to 2025.* in .csproj. For some reason it's launching the Revit 2024 be default. Kindly help to resolve this.  
![Uploading IMG_20260810_164128.jpg…]()
